### PR TITLE
features/caritas-validate-preferences-paper-application-#159245564

### PR DESCRIPTION
**Pull request by fddayan**
https://www.pivotaltracker.com/story/show/159245564

### CODE REVIEW

File: app/javascript/components/applications/application_form/PrimaryApplicantSection.js [hunk 1]
```diff
@@ -16,35 +16,10 @@ let mailingAddressFieldMap = {
...
-const validates = (fun, message) => (value) => {
-  return fun(value) ? null : message
-}
-
-const isOldEnough = (dateOfBirth) => {
-  const years = moment().diff(dateOfBirth,'years')
-  return years >= 18
-}
-
-const isValidDate = (dateOfBirth) => {
```
Comment (akegan) [inline app/javascript/components/applications/application_form/PrimaryApplicantSection.js:22]: looks like this one didn't make it over to your validation page.

File: app/javascript/components/applications/application_form/PrimaryApplicantSection.js [hunk 2]
```diff
@@ -16,35 +16,10 @@ let mailingAddressFieldMap = {
...
-}
-
-const errorClassName = (formApi, field) => {
-  if (formApi.touched[field] || formApi.submitted)
-    return { error: !!formApi.errors[field] }
-  else
-    return null
-}
+const validateError = validate({
+  date_of_birth: validate.isOldEnough("The primary applicant must be 18 years of age or older"),
```
Comment (akegan) [inline app/javascript/components/applications/application_form/PrimaryApplicantSection.js]: missing isValidDate here

File: app/javascript/components/applications/application_form/PrimaryApplicantSection.js [hunk 3]
```diff
@@ -102,22 +68,20 @@ const PrimaryApplicantSection = ({formApi, editValues }) => {
...
-                    type="date"
-                    pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
-                    placeholder="YYYY-MM-DD"
-                    className={classNames(errorClassName(formApi, 'DOB'))} />
-                  <FormError formApi={formApi} field='DOB'/>
-                </div>
+                <Field formApi={formApi} field='DOB' label='DOB' blockNote='- YYYY-MM-DD (required)' >
+                  {(field, classNames) => (
+                    <Text
+                      field="DOB"
```
Comment (akegan) [inline app/javascript/components/applications/application_form/PrimaryApplicantSection.js]: DOB or date_of_birth? it's date_of_birth on line 20 above

File: app/javascript/components/applications/application_form/PrimaryApplicantSection.js [hunk 4]
```diff
@@ -102,22 +68,20 @@ const PrimaryApplicantSection = ({formApi, editValues }) => {
...
-                  <label className='form-label'>Date of Birth</label>
-                  <Text
-                    field="DOB"
-                    type="date"
-                    pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
-                    placeholder="YYYY-MM-DD"
-                    className={classNames(errorClassName(formApi, 'DOB'))} />
-                  <FormError formApi={formApi} field='DOB'/>
-                </div>
+                <Field formApi={formApi} field='DOB' label='DOB' blockNote='- YYYY-MM-DD (required)' >
```
Comment (akegan) [inline app/javascript/components/applications/application_form/PrimaryApplicantSection.js]: this block note is actually incorrect for the way the date is input in chrome with the HTML5 date field:

image

**Review by akegan — DISMISSED** [review]
Just a start. A couple of questions, but most importantly, DOB validation is broken on this PR. We're no longer getting inline errors at all for <18years old or invalid date, either from tabbing away or on save

File: app/javascript/components/applications/application_form/PrimaryApplicantSection.js (hunk 2 above)
Comment (chrisdolendo) [inline app/javascript/components/applications/application_form/PrimaryApplicantSection.js]: I don't think `date_of_birth` and `first_name` is relevant in this version of the code no? I think this should be `DOB` and `firstName`.

File: app/javascript/components/applications/application_form/PrimaryApplicantSection.js (hunk 4 above)
Comment (chrisdolendo) [reply app/javascript/components/applications/application_form/PrimaryApplicantSection.js]: The label on this is old... it should be `Date of Birth (required)`

File: app/javascript/components/applications/application_form/PrimaryApplicantSection.js (hunk 1 above)
Comment (fddayan) [reply app/javascript/components/applications/application_form/PrimaryApplicantSection.js:22]: fixed

**Comment by akegan** [comment]
Date of birth seems good now, but multiple preferences are still wonky.

I tried COP and didn't get any red highlighting until I hit submit, but adding other preferences leads to highlighting before clicking into the fields at all

**Comment by akegan** [comment]
issuewithvalidatepreferences

**Comment by fddayan** [comment]
@akegan as we talked over slack: that problem in the gif image is out of the scope of this story. It's not part of the AC and @slowbot  will have to confirm if that is the desire behavior. It's better if we capture that work in a different story

**Review by akegan — APPROVED** [review]
LGTM now, should we be wrapping every field in a "Field" component going forward?

**Review by chrisdolendo — APPROVED** [review]
:+1: Looks good to me!

File: app/javascript/utils/reactFormUtils.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
+// Shamles copy paste from react-form/redux/utils, Fed
```
Comment (callaghanc) [inline app/javascript/utils/reactFormUtils.js:1]: @fddayan Seems like lodash already offers most of these: https://lodash.com/docs/4.17.10. Can you replace any usage of these with lodash methods where possible, keeping just the ones that are not offered by lodash? These are totally fine, but it's less complicated to stick to one set of util functions as much as possible.

Comment (fddayan) [reply app/javascript/utils/reactFormUtils.js:1]: I tried replacing the lodash methods `isArray`, `isObject`, `flattenDeep` and validations stopped working. There might be subtle differences.I did not dig into why.

Comment (fddayan) [reply app/javascript/utils/reactFormUtils.js:1]: @callaghanc as we talked on slack. I added a NOTE and TODO comment.

Comment (callaghanc) [reply app/javascript/utils/reactFormUtils.js:1]: 🙏

**Review by callaghanc — APPROVED** [review]

**Comment by akegan** [comment]
looks like the PR @chrisdolendo just merged caused merge conflicts on this one

**Pull request merged**